### ISSUE TITLE
Agda allows coinductive records with eta-equality

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#2436]
```agda
record U : Set where
  coinductive
  eta-equality
  field f : U

test : U
test = _

-- Agda loops
```
See also #2197.

**Comment by andreasabel** [comment]
Since we do not allow pattern matching on coinductive constructors, we should not have eta either.

**Comment by andreasabel** [comment]
Outlawed now.  The current syntax does not warn the user he is doing something very dangerous.  If someone needs eta for coinductive records, he could bring it back with an option or pragma.

**Comment by andreasabel** [comment]
Bugger, pushed too early... there is a test case for colists with eta-equality.

**Comment by andreasabel** [comment]
@UlfNorell @Saizan : While `no-eta-equality` is fine as a declaration, I am not so happy about the `eta-equality` declaration in records.  With `eta-equality`, you can make Agda loop.  For this, it looks too regular.  I liked the ETA pragma better, because it looks more like one does something unsafe.  At least in cases where one overrides the positivity checker (recursive unguarded records) or for coinductive records.
I resurrected the ETA pragma from https://github.com/agda/agda/commit/d3b38f393388f40985fcbc5e5d8d18fafc367db4, for now for coinductive records.

**Comment by andreasabel** [comment]
Wishes:
1. Without the `ETA` pragma, Agda should not loop because of infinite eta-expansion.
2. Agda should not silently turn eta on or off (apart from setting it to the default).

Proposal:
1. Use `eta-equality` and `no-eta-equality` to specify the desired behavior of the record (constructor). (One of them might be the default.)
2. Agda switches eta-equality on only after it has been checked positively for safety.
3. The safety test can be overwritten by the `ETA` pragma which turns on eta immediately, delegating the responsibility to the user.
4. The `ETA` pragma is also required for coinductive records.  We might allow pattern matching on coinductive records with eta.

Further comment for discussion: We may want to make `no-eta-equality` the default unless we fix our performance problems.

**Comment by jespercockx** [comment]
I would expect there would be a difference in behaviour between writing `eta-equality`, `no-eta-equality`, or nothing at all: if I write `eta-equality` explicitly and Agda cannot fulfill this demand (e.g. because the record is recursive), I'd expect Agda to tell me so with an error. OTOH, if I didn't write anything then I would be OK with Agda turning on eta-equality if the record is 'safe' and disabling it automatically otherwise.

**Comment by nad** [comment]
> We may want to make `no-eta-equality` the default unless we fix our performance problems.

Please don't do this, it will break existing code.

Do we have any examples of code which becomes *much* slower, or uses *much* more memory, due to eta-equality?

> We might allow pattern matching on coinductive records with eta.

Is this safe? In the past this kind of pattern matching led to issues with subject reduction.

**Comment by UlfNorell** [comment]
> Do we have any examples of code which becomes much slower, or uses much more memory, due to eta-equality?

If you program with pairs it's reasonably easy to run into call-by-name induced exponential behaviour with eta-equality for sigma (although I don't have a concrete example right now).

Whether disabling eta for algebraic (or similar) structures has any considerable performance gains is still not fully investigated.

We are certainly not in a position where making `no-eta-equality` the default is warranted.

**Comment by andreasabel** [comment]
>  > We might allow pattern matching on coinductive records with eta.
> Is this safe? In the past this kind of pattern matching led to issues with subject reduction.

Well, eta-coinductive constructors are no longer lazy, so we can also do pattern matching on them.  By record pattern translation, which is valid for eta-constructors, it vanishes in the next step.

**Comment by UlfNorell** [comment]
Aren't we already doing most of the things in the proposal, aside from pattern matching on coinductive `{-# ETA #-}` constructors? I propose that a new issue be opened if that's something we want to consider. Same for any remaining parts of the proposal that I might have missed. The problem in the OP has most definitely been fixed.
